llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1124,6 +1124,49 @@ void VPlanTransforms::truncateToMinimalBitwidths(
          "some entries in MinBWs haven't been processed");
 }
 
+/// Remove BranchOnCond recipes with constant conditions together with removing
+/// dead edges to their successors. Remove blocks that become dead (no remaining
+/// predecessors())
+static void simplifyCFG(VPlan &Plan) {
+  using namespace llvm::VPlanPatternMatch;
+  SmallVector<VPBasicBlock *> WorkList;
+  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
+           vp_depth_first_deep(Plan.getEntry()))) {
+    VPRecipeBase *Term = VPBB->getTerminator();
+    if (!Term || !match(Term, m_BranchOnCond(m_True())))
+      continue;
+    WorkList.push_back(VPBB);
+  }
+
+  SetVector<VPBasicBlock *> PossiblyDeadBlocks;
+  for (VPBasicBlock *VPBB : WorkList) {
+    VPRecipeBase *Term = VPBB->getTerminator();
+    VPBasicBlock *DeadSucc = cast<VPBasicBlock>(VPBB->getSuccessors()[1]);
+    VPBlockUtils::disconnectBlocks(VPBB, DeadSucc);
+    PossiblyDeadBlocks.insert(DeadSucc);
+    Term->eraseFromParent();
+  }
+  for (VPBasicBlock *VPBB : PossiblyDeadBlocks) {
+    if (VPBB->getNumPredecessors() != 0)
+      continue;
+    // The block doesn't have any predecessors, remove it.
```

Comment:
I guess it could have successors though, which may now also be dead?

Reply:
In the original version this case couldn't happen so far, but it can now it is based on https://github.com/llvm/llvm-project/pull/109975. Rewrote the code to properly handle cases with additional successors and predecessor.